feather/server/src/systems/player_join.rs

```diff
@@ -26,12 +31,114 @@ fn poll_new_players(game: &mut Game, server: &mut Server) -> SysResult {
+
+    let player_data = game.world.load_player_data(client.uuid());
+    let gamemode = player_data
+        .as_ref()
+        .map(|data| Gamemode::from_id(data.gamemode as u8).expect("Unsupported gamemode"))
+        .unwrap_or(server.options.default_gamemode);
+    let previous_gamemode = player_data
+        .as_ref()
+        .map(|data| PreviousGamemode::from_id(data.previous_gamemode as i8))
+        .unwrap_or(PreviousGamemode(None));
+
+    client.send_join_game(gamemode, previous_gamemode);
     client.send_brand();
 
-    let mut builder = game.create_entity_builder(Position::default(), EntityInit::Player);
+    // Abilities
+    let walk_speed = player_data
+        .as_ref()
+        .map(|data| WalkSpeed(data.abilities.walk_speed))
+        .unwrap_or_default();
+    let fly_speed = player_data
+        .as_ref()
+        .map(|data| CreativeFlyingSpeed(data.abilities.fly_speed))
+        .unwrap_or_default();
+    let is_flying = CreativeFlying(
+        player_data
+            .as_ref()
+            .map(|data| data.abilities.flying)
+            .unwrap_or(gamemode == Gamemode::Spectator),
+    );
+    let can_fly = CanCreativeFly(
+        player_data
+            .as_ref()
+            .map(|data| data.abilities.may_fly)
+            .unwrap_or(gamemode == Gamemode::Creative || gamemode == Gamemode::Spectator),
+    );
+    let can_build = CanBuild(
+        player_data
+            .as_ref()
+            .map(|data| data.abilities.may_build)
+            .unwrap_or(gamemode == Gamemode::Creative || gamemode == Gamemode::Survival),
+    );
+    let instabreak = Instabreak(
+        player_data
+            .as_ref()
+            .map(|data| data.abilities.instabuild)
+            .unwrap_or(gamemode == Gamemode::Creative),
+    );
+    let invulnerable = Invulnerable(
+        player_data
+            .as_ref()
+            .map(|data| data.abilities.invulnerable)
+            .unwrap_or(gamemode == Gamemode::Creative || gamemode == Gamemode::Spectator),
+    );
+    client.send_abilities(
+        invulnerable,
+        is_flying,
+        can_fly,
+        instabreak,
+        fly_speed,
+        walk_speed,
+    );
+
+    let hotbar_slot = player_data
+        .as_ref()
+        .map(|data| HotbarSlot::new(data.held_item as usize))
+        .unwrap_or_default();
+    client.set_hotbar_slot(hotbar_slot.get() as u8);
+
+    let mut builder = game.create_entity_builder(
+        player_data
+            .as_ref()
+            .map(|data| Position {
+                x: data.animal.base.position[0],
+                y: data.animal.base.position[1],
+                z: data.animal.base.position[2],
+                yaw: data.animal.base.rotation[0],
+                pitch: data.animal.base.rotation[1],
+            })
+            .unwrap_or_default(),
+        EntityInit::Player,
+    );
 
     let inventory = Inventory::player();
+    if let Ok(data) = player_data.as_ref() {
```

Comment:
This should also be moved to a separate function.
Very minor nitpick, but I would prefer not silently discarding IO errors.